code/controllers/subsystem/blackbox.dm

```diff
@@ -71,112 +48,195 @@ SUBSYSTEM_DEF(blackbox)
 
 /datum/controller/subsystem/blackbox/Shutdown()
 	sealed = FALSE
-	set_val("ahelp_unresolved", GLOB.ahelp_tickets.active_tickets.len)
-
-	var/pda_msg_amt = 0
-	var/rc_msg_amt = 0
-
+	record_feedback("tally", "ahelp_stats", GLOB.ahelp_tickets.active_tickets.len, "unresolved")
 	for (var/obj/machinery/message_server/MS in GLOB.message_servers)
-		if (MS.pda_msgs.len > pda_msg_amt)
-			pda_msg_amt = MS.pda_msgs.len
-		if (MS.rc_msgs.len > rc_msg_amt)
-			rc_msg_amt = MS.rc_msgs.len
-
-	set_details("radio_usage","")
-
-	add_details("radio_usage","COM-[msg_common.len]")
-	add_details("radio_usage","SCI-[msg_science.len]")
-	add_details("radio_usage","HEA-[msg_command.len]")
-	add_details("radio_usage","MED-[msg_medical.len]")
-	add_details("radio_usage","ENG-[msg_engineering.len]")
-	add_details("radio_usage","SEC-[msg_security.len]")
-	add_details("radio_usage","DTH-[msg_deathsquad.len]")
-	add_details("radio_usage","SYN-[msg_syndicate.len]")
-	add_details("radio_usage","SRV-[msg_service.len]")
-	add_details("radio_usage","CAR-[msg_cargo.len]")
-	add_details("radio_usage","OTH-[msg_other.len]")
-	add_details("radio_usage","PDA-[pda_msg_amt]")
-	add_details("radio_usage","RC-[rc_msg_amt]")
+		if (MS.pda_msgs.len)
+			record_feedback("tally", "radio_usage", MS.pda_msgs.len, "PDA")
+		if (MS.rc_msgs.len)
+			record_feedback("tally", "radio_usage", MS.rc_msgs.len, "request console")
+	if(research_levels.len)
+		SSblackbox.record_feedback("associative", "high_research_level", 1, research_levels)
 
 	if (!SSdbcore.Connect())
 		return
 
 	var/list/sqlrowlist = list()
 
 	for (var/datum/feedback_variable/FV in feedback)
-		sqlrowlist += list(list("time" = "Now()", "round_id" = GLOB.round_id, "var_name" =  "'[sanitizeSQL(FV.get_variable())]'", "var_value" = FV.get_value(), "details" = "'[sanitizeSQL(FV.get_details())]'"))
+		var/sqlversion = 1
+		if(FV.key in versions)
+			sqlversion = versions[FV.key]
+		sqlrowlist += list(list("datetime" = "Now()", "round_id" = GLOB.round_id, "key_name" =  "'[sanitizeSQL(FV.key)]'", "version" = "[sqlverion]", "json" = "'[sanitizeSQL(json_encode(FV.json))]'"))
 
 	if (!length(sqlrowlist))
 		return
 
 	SSdbcore.MassInsert(format_table_name("feedback"), sqlrowlist, ignore_errors = TRUE, delayed = TRUE)
 
+/datum/controller/subsystem/blackbox/proc/Seal()
+	if(sealed)
+		return
+	if(IsAdminAdvancedProcCall())
+		message_admins("[key_name_admin(usr)] sealed the blackbox!")
+	log_game("Blackbox sealed[IsAdminAdvancedProcCall() ? " by [key_name(usr)]" : ""].")
+	sealed = TRUE
+
+/datum/controller/subsystem/blackbox/proc/log_research(tech, level)
+	if(!(tech in research_levels) || research_levels[tech] < level)
+		research_levels[tech] = level
 
-/datum/controller/subsystem/blackbox/proc/LogBroadcast(blackbox_msg, freq)
+/datum/controller/subsystem/blackbox/proc/LogBroadcast(freq)
 	if(sealed)
 		return
 	switch(freq)
 		if(1459)

```

Comment:
no const for common?

Reply:
Nope https://github.com/tgstation/tgstation/blob/master/code/game/communications.dm#L158